src/lib/accessibility/FocusZone/FocusTrapZone.tsx

```diff
@@ -0,0 +1,296 @@
+          true,
+        )
+
+    firstFocusableChild && focusAsync(firstFocusableChild)
+  }
+
+  private _onFocusCapture = (ev: React.FocusEvent<HTMLDivElement>) => {
+    this.props.onFocusCapture && this.props.onFocusCapture(ev)
+    if (ev.target !== ev.currentTarget) {
+      // every time focus changes within the trap zone, remember the focused element so that
+      // it can be restored if focus leaves the pane and returns via keystroke (i.e. via a call to this.focus(true))
+      this._previouslyFocusedElementInTrapZone = ev.target as HTMLElement
+    }
+  }
+
+  private _onKeyboardHandler = (ev: React.KeyboardEvent<HTMLDivElement>): void => {
+    this.props.onKeyDown && this.props.onKeyDown(ev)
+
+    if (
+      ev.isDefaultPrevented() ||
+      keyboardKey.getCode(ev) !== keyboardKey.Tab ||
+      !this._root.current
+    ) {
+      return
+    }
+
+    const _firstTabbableChild = getFirstTabbable(
+      this._root.current,
+      this._root.current.firstChild as HTMLElement,
+      true,
+    )
+    const _lastTabbableChild = getLastTabbable(
+      this._root.current,
+      this._root.current.lastChild as HTMLElement,
+      true,
+    )
+
+    if (ev.shiftKey && _firstTabbableChild === ev.target) {
+      focusAsync(_lastTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    } else if (!ev.shiftKey && _lastTabbableChild === ev.target) {
+      focusAsync(_firstTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    }
+  }
+
+  private _forceFocusInTrap = (ev: Event, triggeredElement: HTMLElement) => {
+    if (
+      FocusTrapZone._focusStack.length &&
+      this === FocusTrapZone._focusStack[FocusTrapZone._focusStack.length - 1]
+    ) {
+      if (!this._root.current.contains(triggeredElement)) {
+        this._findElementAndFocusAsync()
+        ev.preventDefault()
+        ev.stopPropagation()
+      }
+    }
+  }
+
+  private _handleOutsideFocus = (ev: FocusEvent): void => {
+    const focusedElement = document.activeElement as HTMLElement
+    focusedElement && this._forceFocusInTrap(ev, focusedElement)
+  }
+
+  private _handleOutsideClick = (ev: MouseEvent): void => {
+    const clickedElement = ev.target as HTMLElement
+    clickedElement && this._forceFocusInTrap(ev, clickedElement)
+  }
+
+  private _subscribeToEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.sub('focus', this._handleOutsideFocus, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+
+    if (!isClickableOutsideFocusTrap) {
+      eventStack.sub('click', this._handleOutsideClick, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+  }
+
+  private _unsubscribeFromEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.unsub('focus', this._handleOutsideFocus, {
```

Comment:
not related to this PR, but this is so prone to accidental mistakes like in the following cases:

#### Case 1
```js
eventStack.sub('focus', this._handleOutsideFocus, { target: window })
...
eventStack.unsub('focus', e => this._handleOutsideFocus(e))
```

#### Case 2
```js
eventStack.sub('focus', e => this._handleOutsideFocus(e))
...
eventStack.unsub('focus', e => this._handleOutsideFocus(e))
```

#### Case 3
```js
private _handleOutsideFocus(e) {
  ...
  this.handle(e)
}
...
eventStack.sub('focus', this._handleOutsideFocus)
```